runners/apex/src/main/java/org/apache/beam/runners/apex/translation/operators/ApexParDoOperator.java

```diff
@@ -165,8 +171,16 @@ public ApexParDoOperator(
       Coder<?> keyCoder = StringUtf8Coder.of();
       this.currentKeyStateInternals = new StateInternalsProxy<>(
           stateBackend.newStateInternalsFactory(keyCoder));
+    } else {
+      DoFnSignature signature = DoFnSignatures.getSignature(doFn.getClass());
+      if (signature.stateDeclarations().size() > 0) {
```

Comment:
<!--new_thread; commit:ee392026f52a00eb204c6b3bd8498eb9e82c5cc3; resolved:1-->
In review, someone suggested I add `signature.usesState()` to encapsulate this check. You could use it or not, I don't care so much.